Rewrites the discovery mechanism to circumvent the non-blocking natur…

markus-orblom opened the pull request:
…e of the UPNP SSDP library.

Unintentioned exits before responses from the UPNP SSDP search has been received will thus no longer happen.

This has been implemented through the addition of two classes Discoverer and Device. The Discoverer
class contains the main business logic to make the UPNP SSDP search blocking (by timeouts), while the
Device class is just a basic data storage for all interesting information for a device, which currently
only is the address and what status code it responded with.
The Discoverer class is implemented as an event dispatcher. For every response from the UPNP SSDP search it
will itself emit its own discovery event with the found device and a list of all devices. The duration of
the discovery is customizable. It defaults to 5000 ms, but can be set to anything. If it is set to a value
<= 0 the discovery will continue indefinitely. When the discovery phase ends it stops the UPNP SSDP listener

The heos module now exposes the Discoverer class, the discover function and a simpler findAny function. The
latter two functions can be used if the use case is sufficiently simple to not require the flexibility
offered by the Discoverer class.

The discover function has had its footprint modified and it now takes a callback as an argument. The
callback us used for the callback for the discovery event emitted by the Discoverer object during
its discovery.

The findAny function resolves its promise with the first device found from the discoverer.

juliuscc commented:
I really like the concept of extending the EventEmitter class. This is exactly the kind of solution that I would imagine. I have a couple of questions though.

1. It does not seem as this follows the style guide previously used. To fix that just do
```bash
npm install && npm run prettify
```
to make sure that all files follow the standard. If you have ever done ```npm install``` before git committing any commits, the files will automatically be prettified according to the style guide.

2. On [line 48](https://github.com/markus-orblom/heos-api/blob/50dfee855ef0b12e3737a02240bb61041822cd20/lib/connect.js#L48) the timeout is set to either 0, a negative number or anything that is not a number. What will happen in the first two cases is that the interval will be set to [10 ms](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setInterval). If timeout is not a number we can disregard, cause that should be undefined behaviour. As 10ms is very low, we could as likely just set the timeout to a very high number. Even though the function is empty, the function is run very often. I suggest we can hardcode this to maybe 10000ms or more.

3. Is it necessary that Device is a class? It could be just an object.

4. The promise on [line 62](https://github.com/markus-orblom/heos-api/blob/50dfee855ef0b12e3737a02240bb61041822cd20/lib/connect.js#L62) is never resolved or rejected. I think the Promise could be omitted, and nothing is returned.

5. If the "stop-sequence" on [line 50](https://github.com/markus-orblom/heos-api/blob/50dfee855ef0b12e3737a02240bb61041822cd20/lib/connect.js#L50) is put in it's own function (possibly called ```stop()```), then the discovery could be stopped. I see this being useful both if the timeout is infinite where the search could be manually stopped, or in ```findAny()``` where the search should be stopped after the first resolve.

I really like the changes made, and am really thankful for the help. 🙇 👏  I haven't really had time to focus on this. If you do not have time to commit the changes I might be able to fix this in the weekend.

P.S. If you introduced the ```Device```-class for type-recognition, you might be interested in my typescript-branch. I have started working on migrating the package to typescript for more robust code, and for type definitions.

juliuscc commented:
I fixed issue # 1

juliuscc commented:
I have fixed all issues I had. Please try it for yourself and tell me if you like it.

juliuscc commented:
The moment you say that it's good i will merge the pull request

juliuscc commented:
Yes I'll add a helper function to do the transformation ssdp device -> potential heos device. It could make the code easier to read. The device should later be an interface in typescript and not a class as a class is superfluous.

I like the idea of testing this automatically. I do not think the node-ssdp library should be tested though. If we set up both a server and a client it feels as if that is what we are testing. It might also be difficult to reliably test on Travis. What we should do is simulate the server with dependency injection. Basically what that is is that we in testing mock the node-ssdp package and present our own package that mocks behaviour we want to test.

juliuscc commented:
Now I've done the updates. I think that the extended testing could be a separate issue. This works perfectly for me as it is now so I will not prioritize testing of this. If you feel up for the task you are very welcome to apply testing on a new pull request. 😄

markus-orblom commented:
I have been a bit busy lately. I hope I will get some time to test this now. I added a small improvement. The generation of the no device found error is now moved into a function and is not duplicated.

juliuscc commented:
Really good catch! I am in no rush. Make sure that it works for you and then I will merge it in. It works for me already.

juliuscc commented:
Perfect! Thanks for the help!
